src/Sentry/Internal/MemoryMonitor.cs

```diff
@@ -0,0 +1,172 @@
+            ? $"dotnet {bundledToolPath} {arguments}"
+            : $"dotnet-gcdump {arguments}";
+
+        _options.LogDebug($"Starting process: {command}");
+        using var process = new Process();
+        process.StartInfo = new ProcessStartInfo
+        {
+            FileName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "cmd.exe" : "/bin/bash",
+            Arguments = $"-c \"{command}\"",
+            RedirectStandardOutput = true,
+            RedirectStandardError = false,
+            UseShellExecute = false,
+            CreateNoWindow = true,
+        };
+        process.Start();
+        while (!process.StandardOutput.EndOfStream)
+        {
+            if (process.StandardOutput.ReadLine() is { } line)
+            {
+                _options.LogDebug($"gcdump: {line}");
+            }
+        }
+#if NET8_0_OR_GREATER
+        process.WaitForExit(TimeSpan.FromSeconds(5));
+#else
+        process.WaitForExit(5000);
+#endif
+
+        if (!_options.FileSystem.FileExists(dumpFile))
+        {
+            // if this happens, hopefully there would be more information in the standard output from the process above
+            _options.LogError("Unexpected error creating memory dump. Use debug-level to see output of dotnet-gcdump.");
+        }
+
+        _onDumpCollected(dumpFile);
+    }
+
+    internal string? TryGetDumpLocation()
+    {
+        try
+        {
+            var rootPath = _options.CacheDirectoryPath ??
+                           Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            var directoryPath = Path.Combine(rootPath, "Sentry", _options.Dsn!.GetHashString());
+            var fileSystem = _options.FileSystem;
+
+            if (!fileSystem.CreateDirectory(directoryPath))
+            {
+                _options.LogWarning("Failed to create a directory for memory dump ({0}).", directoryPath);
+                return null;
+            }
+            _options.LogDebug("Created directory for heap dump ({0}).", directoryPath);
+
+            var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
+            var processId = Environment.ProcessId;
+            var filePath = Path.Combine(directoryPath, $"{timestamp}_{processId}.gcdump");
+            if (fileSystem.FileExists(filePath))
+            {
+                _options.LogWarning("Duplicate dump file detected.");
+                return null;
+            }
+
+            return filePath;
+        }
+        // If there's no write permission or the platform doesn't support this, we handle simply log and bug out
+        catch (Exception ex)
+        {
+            _options.LogError(ex, "Failed to resolve appropriate memory dump location.");
+            return null;
+        }
+    }
+
+    public void Dispose()
+    {
+        // Important no exceptions can be thrown from this method as it's called when disposing the Hub
+        _cancellationTokenSource.Cancel();
```

Comment:
CancellationTokenSource has to be disposed too (if you combine tokens with it specially, I recall a memory leak at NRGS because of that :) 

took a while to track down. Ironically I imagine this feature would help with that lol.

Reply:
Ah, good to know... but yes, it would be ironic if the tool we used to help detect memory leaks had a memory leak.